module/os/linux/zfs/zvol_os.c

```diff
@@ -83,6 +117,108 @@ zv_request_task_free(zv_request_task_t *task)
 	kmem_free(task, sizeof (*task));
 }
 
+#ifdef HAVE_BLK_MQ
+/*
+ * This is our blk-mq workqueue callback function.  It's here that
+ * we process all the BIOs in a request.
+ */
+static void zvol_mq_work_func(struct work_struct *work)
+{
+	zv_work_t *zv_work;
+	struct bio *bio = NULL;
+	struct request *rq;
+	zvol_state_t *zv;
+	blk_status_t res = BLK_STS_OK;
+
+	zv_work = container_of(work, zv_work_t, work);
+	rq = zv_work->rq;
+	zv = rq->q->queuedata;
+
+	/* Tell the kernel that we are starting to process this request */
+	blk_mq_start_request(rq);
+
+	if (blk_rq_is_passthrough(rq)) {
+		/* Skip non filesystem request */
+		blk_mq_end_request(rq, BLK_STS_IOERR);
+		goto out;
+	}
+
+	/* Execute the BIOs in this request. */
+	__rq_for_each_bio(bio, rq) {
+		zvol_request_impl(zv, bio, 1);
+
+		/* Did this BIO cause an error?  If so, stop the request */
+#ifdef HAVE_BIO_BI_STATUS
+		res = bio->bi_status;
+#else
+		res = bio->bi_error;
+#endif
+		if (res != BLK_STS_OK) {
+			/* Got an error */
+			break;
+		}
+	}
+
+	/* All done */
+	blk_mq_end_request(rq, res);
+
+out:
+	kmem_cache_free(blk_mq_cache, zv_work);
+}
+
+/*
+ * This is called when a new block multiqueue request comes in.  A request
+ * contains one or more BIOs.  This function is run from an atomic context.
+ */
+static blk_status_t zvol_mq_queue_rq(struct blk_mq_hw_ctx *hctx,
+    const struct blk_mq_queue_data *bd)
+{
+	struct request *rq = bd->rq;
+	zv_work_t *zv_work;
+
+	/*
+	 * Add the request to our workqueue for processing later since we're
+	 * currently in an atomic context and our bio processing code can sleep.
+	 */
+	zv_work = kmem_cache_alloc(blk_mq_cache, KM_NOSLEEP);
+	if (!zv_work) {
+		/* Try again, maybe memory will free up */
+		return (BLK_STS_AGAIN);
+	}
+	INIT_WORK(&zv_work->work, zvol_mq_work_func);
+	zv_work->rq = rq;
+	schedule_work(&zv_work->work);
+
+	/* Acknowledge to the kernel that we got this request */
+	return (BLK_STS_OK);
+}
+
+static struct blk_mq_ops my_queue_ops = {
+	.queue_rq = zvol_mq_queue_rq,
+};
+
+/* Initialize our blk-mq struct */
+static int zvol_blk_mq_alloc_tag_set(zvol_state_t *zv)
+{
+	int err;
+	struct zvol_state_os *zso = zv->zv_zso;
+
+	memset(&zso->tag_set, 0, sizeof (zso->tag_set));
+
+	/* Initialize tag set. */
+	zso->tag_set.ops = &my_queue_ops;
+	zso->tag_set.nr_hw_queues = zvol_actual_threads;
+	zso->tag_set.queue_depth = zvol_actual_blk_mq_queue_depth;
+	zso->tag_set.numa_node = NUMA_NO_NODE;
+	zso->tag_set.cmd_size = 0;
+	zso->tag_set.flags = BLK_MQ_F_SHOULD_MERGE;
+	zso->tag_set.driver_data = zv;
+	err = blk_mq_alloc_tag_set(&zso->tag_set);
+
+	return (err);
```

Comment:
Looks like you could just `return (blk_mq_alloc_tag_set(&zso->tag_set));` and not bother with `err` at all.